libraries/AP_Camera/AP_Camera.h

```diff
@@ -10,173 +10,141 @@
 #include <AP_Logger/LogStructure.h>
 #include <AP_Param/AP_Param.h>
 #include <GCS_MAVLink/GCS_MAVLink.h>
+#include "AP_Camera_Params.h"
 
-#define AP_CAMERA_TRIGGER_DEFAULT_DURATION  10      // default duration servo or relay is held open in 10ths of a second (i.e. 10 = 1 second)
+#define AP_CAMERA_MAX_INSTANCES             2       // maximum number of camera backends
 
-#define AP_CAMERA_SERVO_ON_PWM              1300    // default PWM value to move servo to when shutter is activated
-#define AP_CAMERA_SERVO_OFF_PWM             1100    // default PWM value to move servo to when shutter is deactivated
-
-#define AP_CAMERA_FEEDBACK_DEFAULT_FEEDBACK_PIN -1  // default is to not use camera feedback pin
+// declare backend classes
+class AP_Camera_Backend;
+class AP_Camera_Servo;
+class AP_Camera_Relay;
+class AP_Camera_SoloGimbal;
+class AP_Camera_Mount;
 
 /// @class	Camera
 /// @brief	Object managing a Photo or video camera
 class AP_Camera {
 
+    // declare backends as friends
+    friend class AP_Camera_Backend;
+    friend class AP_Camera_Servo;
+    friend class AP_Camera_Relay;
+    friend class AP_Camera_SoloGimbal;
+    friend class AP_Camera_Mount;
+    friend class AP_Camera_MAVLink;
+
 public:
-    AP_Camera(uint32_t _log_camera_bit)
-        : log_camera_bit(_log_camera_bit)
-    {
-        AP_Param::setup_object_defaults(this, var_info);
-        _singleton = this;
-    }
+
+    // constructor
+    AP_Camera(uint32_t _log_camera_bit);
 
     /* Do not allow copies */
     CLASS_NO_COPY(AP_Camera);
 
     // get singleton instance
-    static AP_Camera *get_singleton()
-    {
-        return _singleton;
-    }
+    static AP_Camera *get_singleton() { return _singleton; }
+
+    // enums
+    enum class CameraType {
+        NONE = 0,           // None
+        SERVO = 1,          // Servo/PWM controlled camera
+        RELAY = 2,          // Relay controlled camera
+        SOLOGIMBAL = 3,     // GoPro in Solo gimbal
+        MOUNT = 4,          // Mount library implements camera
+        MAVLINK = 5,        // MAVLink enabled camera
+    };
+
+    // returns camera type of the given instance
+    CameraType get_type(uint8_t instance) const;
+
+    // detect and initialise backends
+    void init();
+
+    // update - to be called periodically at 50Hz
+    void update();
 
     // MAVLink methods
-    void            handle_message(mavlink_channel_t chan,
-                                   const mavlink_message_t &msg);
-    void            send_feedback(mavlink_channel_t chan) const;
+    void handle_message(mavlink_channel_t chan, const mavlink_message_t &msg);
+    void send_feedback(mavlink_channel_t chan) const;
+
+    // configure camera
+    void configure(float shooting_mode, float shutter_speed, float aperture, float ISO, float exposure_type, float cmd_id, float engine_cutoff_time);
```

Comment:
Would it be better to pass configure and control arguments as a structure instead?

Reply:
txs for the review.  It might be.. but for this PR I tried to keep the scope limited to getting the frontend/backend split completed without adding any many additional enhancements that could slow its merge to master.